src/rgw/rgw_keystone_scope.h

```diff
@@ -0,0 +1,205 @@
+   * Domains provide namespace isolation in Keystone.
+   */
+  struct domain_t {
+    std::string id;
+    std::string name;
+
+    void encode(bufferlist &bl) const {
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      ENCODE_FINISH(bl);
+    }
+    void decode(bufferlist::const_iterator &p) {
+      DECODE_START(1, p);
+      decode(id, p);
+      decode(name, p);
+      DECODE_FINISH(p);
+    }
+  };
+
+  /**
+   * Keystone project (tenant) information.
+   * Projects are the primary authorization scope in Keystone.
+   */
+  struct project_t {
+    std::string id;
+    std::string name;
+    domain_t domain;
+
+    void encode(bufferlist &bl) const {
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      domain.encode(bl);
+      ENCODE_FINISH(bl);
+    }
+    void decode(bufferlist::const_iterator &p) {
+      DECODE_START(1, p);
+      decode(id, p);
+      decode(name, p);
+      domain.decode(p);
+      DECODE_FINISH(p);
+    }
+  };
+
+  /**
+   * Keystone user information.
+   * Optional based on rgw_keystone_scope_include_user configuration.
+   */
+  struct user_t {
+    std::string id;
+    std::string name;
+    domain_t domain;
+
+    void encode(bufferlist &bl) const {
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      domain.encode(bl);
+      ENCODE_FINISH(bl);
+    }
+    void decode(bufferlist::const_iterator &p) {
+      DECODE_START(1, p);
+      decode(id, p);
+      decode(name, p);
+      domain.decode(p);
+      DECODE_FINISH(p);
+    }
+  };
+
+  /**
+   * Keystone application credential information.
+   * Only present when authentication used application credentials.
+   */
+  struct app_cred_t {
+    std::string id;
+    std::string name;
+    bool restricted;
+
+    void encode(bufferlist &bl) const {
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      encode(restricted, bl);
+      ENCODE_FINISH(bl);
+    }
+    void decode(bufferlist::const_iterator &p) {
+      DECODE_START(1, p);
+      decode(id, p);
+      decode(name, p);
+      decode(restricted, p);
+      DECODE_FINISH(p);
+    }
+  };
+
+  // Fields
+  project_t project;                        // Always present
+  std::optional<user_t> user;               // Optional (controlled by include_user config)
+  std::vector<std::string> roles;           // May be empty (controlled by include_roles config)
+  std::optional<app_cred_t> app_cred;       // Optional (present only for app cred auth)
+
+  // Serialization for RADOS backend
+  void encode(bufferlist &bl) const {
+    ENCODE_START(1, 1, bl);
+    project.encode(bl);
+
+    __u8 has_user = user.has_value();
+    encode(has_user, bl);
+    if (has_user) {
+      user->encode(bl);
+    }
+
+    encode(roles, bl);
+
+    __u8 has_app_cred = app_cred.has_value();
+    encode(has_app_cred, bl);
+    if (has_app_cred) {
+      app_cred->encode(bl);
+    }
```

Comment:
note that include/encoding.h already supports encode/decode for std::optional, so you should be able to `encode(user, bl)` and `encode(app_cred, bl)` directly

same for `rgw_log_entry::keystone_scope`